File: KeePassLib/Cryptography/KeyDerivation/AesKdf.cs
/*
  KeePass Password Safe - The Open-Source Password Manager
  Copyright (C) 2003-2025 Dominik Reichl <dominik.reichl@t-online.de>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;
using System.Threading;

#if KeePassUAP
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Engines;
using Org.BouncyCastle.Crypto.Parameters;
#else
using System.Security.Cryptography;
#endif

using KeePassLib.Native;
using KeePassLib.Utility;

namespace KeePassLib.Cryptography.KeyDerivation
{
	public sealed partial class AesKdf : KdfEngine
	{
		private static readonly PwUuid g_uuid = new PwUuid(new byte[] {
			0xC9, 0xD9, 0xF3, 0x9A, 0x62, 0x8A, 0x44, 0x60,
			0xBF, 0x74, 0x0D, 0x08, 0xC1, 0x8A, 0x4F, 0xEA });

		public static readonly string ParamSeed = "S"; // Byte[32]
		public static readonly string ParamRounds = "R"; // UInt64

		private const int AkcBufferBlocks = 8192;

		public override PwUuid Uuid
		{
			get { return g_uuid; }
		}

		public override string Name
		{
			get { return "AES-KDF"; }
		}

		public AesKdf()
		{
		}

		public override KdfParameters GetDefaultParameters()
		{
			KdfParameters p = base.GetDefaultParameters();
			p.SetUInt64(ParamRounds, PwDefs.DefaultKeyEncryptionRounds);
			return p;
		}

		public override bool AreParametersWeak(KdfParameters p)
		{
			if(p == null) { Debug.Assert(false); return false; }

			ulong uRounds = p.GetUInt64(ParamRounds, ulong.MaxValue);
			return (uRounds < PwDefs.DefaultKeyEncryptionRounds);
		}

		public override void Randomize(KdfParameters p)
		{
			if(p == null) { Debug.Assert(false); return; }
			Debug.Assert(g_uuid.Equals(p.KdfUuid));

			byte[] pbSeed = CryptoRandom.Instance.GetRandomBytes(32);
			p.SetByteArray(ParamSeed, pbSeed);
		}

		public override byte[] Transform(byte[] pbMsg, KdfParameters p)
		{
			if(pbMsg == null) throw new ArgumentNullException("pbMsg");
			if(p == null) throw new ArgumentNullException("p");

			Type tRounds = p.GetTypeOf(ParamRounds);
			if(tRounds == null) throw new ArgumentNullException("p.Rounds");
			if(tRounds != typeof(ulong)) throw new ArgumentOutOfRangeException("p.Rounds");
			ulong uRounds = p.GetUInt64(ParamRounds, 0);

			byte[] pbSeed = p.GetByteArray(ParamSeed);
			if(pbSeed == null) throw new ArgumentNullException("p.Seed");

			if(pbMsg.Length != 32)
			{
				Debug.Assert(false);
				pbMsg = CryptoUtil.HashSha256(pbMsg);
			}

			if(pbSeed.Length != 32)
			{
				Debug.Assert(false);
				pbSeed = CryptoUtil.HashSha256(pbSeed);
			}

			return TransformKey(pbMsg, pbSeed, uRounds);
		}

		private static byte[] TransformKey(byte[] pbData32, byte[] pbSeed32,
			ulong uRounds)
		{
			if(pbData32 == null) throw new ArgumentNullException("pbData32");
			if(pbData32.Length != 32) throw new ArgumentOutOfRangeException("pbData32");
			if(pbSeed32 == null) throw new ArgumentNullException("pbSeed32");
			if(pbSeed32.Length != 32) throw new ArgumentOutOfRangeException("pbSeed32");

			byte[] pbTrf32 = new byte[32];
			Array.Copy(pbData32, pbTrf32, 32);

			try
			{
				if(NativeLib.TransformKey256(pbTrf32, pbSeed32, uRounds))
					return CryptoUtil.HashSha256(pbTrf32);

				if(TransformKeyGCrypt(pbTrf32, pbSeed32, uRounds))
					return CryptoUtil.HashSha256(pbTrf32);

				TransformKeyManaged(pbTrf32, pbSeed32, uRounds);
				return CryptoUtil.HashSha256(pbTrf32);
			}
			finally { MemUtil.ZeroByteArray(pbTrf32); }
		}

		internal static void TransformKeyManaged(byte[] pbData32, byte[] pbSeed32,
			ulong uRounds)
		{
			if(uRounds == 0) return;

#if KeePassUAP
			KeyParameter kp = new KeyParameter(pbSeed32);
			AesEngine aes = new AesEngine();
			aes.Init(true, kp);

			for(ulong u = 0; u < uRounds; ++u)
			{
				aes.ProcessBlock(pbData32, 0, pbData32, 0);
				aes.ProcessBlock(pbData32, 16, pbData32, 16);
			}

			aes.Reset();
#else
			byte[] pbDataL = null, pbDataR = null;
			try
			{
				pbDataL = MemUtil.Mid(pbData32, 0, 16);
				pbDataR = MemUtil.Mid(pbData32, 16, 16);

				Exception exL = null, exR = null;

				Thread thL = new Thread(new ThreadStart(() => { exL =
					TransformKeyManagedHalf(pbDataL, pbSeed32, uRounds); }));
				Thread thR = new Thread(new ThreadStart(() => { exR =
					TransformKeyManagedHalf(pbDataR, pbSeed32, uRounds); }));
				thL.Start();
				thR.Start();
				thL.Join();
				thR.Join();

				if(exL != null) throw new Exception(null, exL);
				if(exR != null) throw new Exception(null, exR);

				Array.Copy(pbDataL, 0, pbData32, 0, 16);
				Array.Copy(pbDataR, 0, pbData32, 16, 16);
			}
			finally
			{
				MemUtil.ZeroByteArray(pbDataL);
				MemUtil.ZeroByteArray(pbDataR);
			}
#endif
		}

		private static Exception TransformKeyManagedHalf(byte[] pbData16,
			byte[] pbSeed32, ulong uRounds)
		{
			byte[] pbBuf = null;
			try
			{
				byte[] pbZero = new byte[AkcBufferBlocks * 16];
				pbBuf = new byte[AkcBufferBlocks * 16];

				using(SymmetricAlgorithm a = CryptoUtil.CreateAes(256, CipherMode.CBC,
					PaddingMode.None))
				{
					using(ICryptoTransform t = a.CreateEncryptor(pbSeed32, pbData16))
					{
						while(true)
						{
							ulong cBlocks = Math.Min(uRounds, (ulong)AkcBufferBlocks);
							int cBytes = (int)cBlocks << 4;

							t.TransformBlock(pbZero, 0, cBytes, pbBuf, 0);

							uRounds -= cBlocks;
							if(uRounds == 0)
							{
								if(cBytes == 0) { Debug.Assert(false); }
								else Array.Copy(pbBuf, cBytes - 16, pbData16, 0, 16);

								break;
							}
						}
					}
				}
			}
			catch(Exception ex) { Debug.Assert(false); return ex; }
			finally { MemUtil.ZeroByteArray(pbBuf); }

			return null;
		}

		public override KdfParameters GetBestParameters(uint uMilliseconds)
		{
			ulong uRounds;

			if(NativeLib.TransformKeyBenchmark256(uMilliseconds, out uRounds)) { }
			else if(TransformKeyGCryptBenchmark(uMilliseconds, out uRounds)) { }
			else
			{
				Exception exL = null, exR = null;
				ulong uRoundsL = 0, uRoundsR = 0;

				Thread thL = new Thread(new ThreadStart(() => { exL =
					TransformKeyManagedBenchmarkHalf(uMilliseconds, out uRoundsL); }));
				Thread thR = new Thread(new ThreadStart(() => { exR =
					TransformKeyManagedBenchmarkHalf(uMilliseconds, out uRoundsR); }));
				thL.Start();
				thR.Start();
				thL.Join();
				thR.Join();

				if(exL != null) throw new Exception(null, exL);
				if(exR != null) throw new Exception(null, exR);

				uRounds = (uRoundsL >> 1) + (uRoundsR >> 1);
			}

			KdfParameters p = GetDefaultParameters();
			if(uRounds != 0) p.SetUInt64(ParamRounds, uRounds);
			else { Debug.Assert(false); }
			return p;
		}

		private static Exception TransformKeyManagedBenchmarkHalf(uint uMilliseconds,
			out ulong uRounds)
		{
			uRounds = 0;

			try
			{
				const ulong cBlocks = AkcBufferBlocks;
				const int cBytes = checked((int)cBlocks * 16);

				byte[] pbData16 = new byte[16];
				byte[] pbSeed32 = new byte[32];
				byte[] pbZero = new byte[cBytes];
				byte[] pbBuf = new byte[cBytes];

				Random r = CryptoRandom.NewWeakRandom();
				r.NextBytes(pbData16);
				r.NextBytes(pbSeed32);

				using(SymmetricAlgorithm a = CryptoUtil.CreateAes(256, CipherMode.CBC,
					PaddingMode.None))
				{
					using(ICryptoTransform t = a.CreateEncryptor(pbSeed32, pbData16))
					{
						int tStart = Environment.TickCount;
						while((uint)(Environment.TickCount - tStart) <= uMilliseconds)
						{
							t.TransformBlock(pbZero, 0, cBytes, pbBuf, 0);

							uRounds += cBlocks;
							if(uRounds < cBlocks) // Overflow check
							{
								uRounds = ulong.MaxValue;
								break;
							}
						}
					}
				}
			}
			catch(Exception ex) { Debug.Assert(false); return ex; }

			return null;
		}
	}
}

File: KeePassLibN/Cryptography/AesKdf/AesKdf.cpp
/*
  KeePassLibN
  Copyright (C) 2003-2025 Dominik Reichl <dominik.reichl@t-online.de>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "AesKdf.h"
#include "../../Utility/AppUtil.h"

#include <TimeApi.h>

#define AKC_FAIL { assert(false); return; }

constexpr size_t AkcBufferBlocks = 8192;
constexpr size_t AkcBufferSize = AkcBufferBlocks * 16;

CAesKdf::CAesKdf(const uint8_t* pbSeed32) :
	m_hAlg(NULL), m_hKey(NULL), m_bReady(false)
{
	if(pbSeed32 == nullptr) { assert(false); return; }

	if(!BCRYPT_SUCCESS(BCryptOpenAlgorithmProvider(&m_hAlg, BCRYPT_AES_ALGORITHM,
		nullptr, 0)))
		AKC_FAIL;
	if(m_hAlg == NULL) { BCryptCloseAlgorithmProvider(m_hAlg, 0); AKC_FAIL; }

	DWORD cbKeyObj = 0;
	ULONG cbResult = 0;
	if(!BCRYPT_SUCCESS(BCryptGetProperty(m_hAlg, BCRYPT_OBJECT_LENGTH,
		(PUCHAR)&cbKeyObj, sizeof(DWORD), &cbResult, 0)) || (cbKeyObj == 0))
		AKC_FAIL;

	m_upKeyObj.reset(new CAlignedBuffer(cbKeyObj, 16, true, true));
	if((m_upKeyObj.get() == nullptr) || (m_upKeyObj->Data() == nullptr)) AKC_FAIL;

	constexpr ULONG cbKeyBlobHeader = sizeof(BCRYPT_KEY_DATA_BLOB_HEADER);
	constexpr ULONG cbKeyBlob = cbKeyBlobHeader + 32;
	uint8_t vKeyBlob[cbKeyBlob] = { 0 };
	BCRYPT_KEY_DATA_BLOB_HEADER* pKeyBlob = (BCRYPT_KEY_DATA_BLOB_HEADER*)vKeyBlob;
	pKeyBlob->dwMagic = BCRYPT_KEY_DATA_BLOB_MAGIC;
	pKeyBlob->dwVersion = BCRYPT_KEY_DATA_BLOB_VERSION1;
	pKeyBlob->cbKeyData = 32;
	memcpy(vKeyBlob + cbKeyBlobHeader, pbSeed32, 32);

	const NTSTATUS s = BCryptImportKey(m_hAlg, NULL, BCRYPT_KEY_DATA_BLOB,
		&m_hKey, m_upKeyObj->Data(), cbKeyObj, vKeyBlob, cbKeyBlob, 0);
	SecureZeroMemory(vKeyBlob, cbKeyBlob);
	if(!BCRYPT_SUCCESS(s)) AKC_FAIL;

	LPCWSTR lpwCbc = BCRYPT_CHAIN_MODE_CBC;
	if(!BCRYPT_SUCCESS(BCryptSetProperty(m_hAlg, BCRYPT_CHAINING_MODE,
		(PUCHAR)lpwCbc, (ULONG)((wcslen(lpwCbc) + 1) * sizeof(WCHAR)), 0)))
		AKC_FAIL;

#ifdef _DEBUG
	DWORD cbKey = 0;
	assert(BCRYPT_SUCCESS(BCryptGetProperty(m_hKey, BCRYPT_KEY_LENGTH,
		(PUCHAR)&cbKey, sizeof(DWORD), &cbResult, 0)) && (cbKey == 256));

	BCRYPT_ALG_HANDLE hAlgRef = NULL;
	assert(BCRYPT_SUCCESS(BCryptGetProperty(m_hKey, BCRYPT_PROVIDER_HANDLE,
		(PUCHAR)&hAlgRef, sizeof(BCRYPT_ALG_HANDLE), &cbResult, 0)) &&
		(hAlgRef == m_hAlg));
#endif

	m_upIV.reset(new CAlignedBuffer(16, 16, true, true));
	if((m_upIV.get() == nullptr) || (m_upIV->Data() == nullptr)) AKC_FAIL;

	m_upZero.reset(new CAlignedBuffer(AkcBufferSize, 16, true, false));
	if((m_upZero.get() == nullptr) || (m_upZero->Data() == nullptr)) AKC_FAIL;

	m_upBuf.reset(new CAlignedBuffer(AkcBufferSize, 16, false, true));
	if((m_upBuf.get() == nullptr) || (m_upBuf->Data() == nullptr)) AKC_FAIL;

	m_bReady = true;
}

CAesKdf::~CAesKdf()
{
	if(m_hKey != NULL) { BCryptDestroyKey(m_hKey); m_hKey = NULL; }
	if(m_hAlg != NULL) { BCryptCloseAlgorithmProvider(m_hAlg, 0); m_hAlg = NULL; }
}

bool CAesKdf::TransformHalf(uint8_t* pbData16, uint64_t uRounds)
{
	if(!m_bReady) return false;

	uint8_t* pbIV = m_upIV->Data();
	if(pbData16 != nullptr) memcpy(pbIV, pbData16, 16);

	uint8_t* pbZero = m_upZero->Data();
	uint8_t* pbBuf = m_upBuf->Data();

	const bool bWine = AuxIsWine();

	while(uRounds != 0)
	{
		const uint64_t r = min(uRounds, (uint64_t)AkcBufferBlocks);
		const ULONG cb = (ULONG)r << 4;
		ULONG cbResult = 0;

		if(!BCRYPT_SUCCESS(BCryptEncrypt(m_hKey, pbZero, cb, nullptr, pbIV, 16,
			pbBuf, cb, &cbResult, 0)))
		{
			assert(false);
			return false;
		}

		assert(*(uint64_t*)pbZero == 0);
		assert(memcmp(pbIV, pbBuf + (cb - 16), 16) == 0);
		assert(cbResult == cb);

		// Workaround for https://bugs.winehq.org/show_bug.cgi?id=52457
		if(bWine) memcpy(pbIV, pbBuf + (cb - 16), 16);

		uRounds -= r;
	}

	if(pbData16 != nullptr) memcpy(pbData16, pbIV, 16);
	return true;
}

KPL_API BOOL AesKdfTransformHalf(uint8_t* pbData16, const uint8_t* pbSeed32,
	uint64_t uRounds)
{
	CAesKdf kdf(pbSeed32);
	return (kdf.TransformHalf(pbData16, uRounds) ? TRUE : FALSE);
}

KPL_API uint64_t AesKdfTransformBenchmarkHalf(uint32_t uMilliseconds)
{
	constexpr uint64_t cBlocks = AkcBufferBlocks;
	constexpr uint8_t vSeed32[32] = { 0 };

	CAesKdf kdf(vSeed32);
	uint64_t r = 0;

	const DWORD tStart = timeGetTime();
	while((timeGetTime() - tStart) <= uMilliseconds)
	{
		if(!kdf.TransformHalf(nullptr, cBlocks)) return 0;

		r += cBlocks;
		if(r < cBlocks) { r = UINT64_MAX; break; }
	}

	return r;
}
